From cfe43d895a3cd9b053f6b4fbb2c419f2c5065257 Mon Sep 17 00:00:00 2001
From: Tomas Orsava <torsava@redhat.com>
Date: Tue, 10 Jan 2017 11:05:27 +0100
Subject: [PATCH] Removes the "-g" option from "pydoc"

---
 Lib/pydoc.py | 11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/Lib/pydoc.py b/Lib/pydoc.py
index b4b190f..183027d 100755
--- a/Lib/pydoc.py
+++ b/Lib/pydoc.py
@@ -20,9 +20,6 @@ Run "pydoc -p <port>" to start an HTTP server on a given port on the
 local machine to generate documentation web pages.  Port number 0 can be
 used to get an arbitrary unused port.
 
-For platforms without a command line, "pydoc -g" starts the HTTP server
-and also pops up a little window for controlling it.
-
 Run "pydoc -w <name>" to write out the HTML documentation for a module
 to a file named "<name>.html".
 
@@ -2341,9 +2338,6 @@ def cli():
         writing = 0
 
         for opt, val in opts:
-            if opt == '-g':
-                gui()
-                return
             if opt == '-k':
                 apropos(val)
                 return
@@ -2398,13 +2392,10 @@ def cli():
     Start an HTTP server on the given port on the local machine.  Port
     number 0 can be used to get an arbitrary unused port.
 
-%s -g
-    Pop up a graphical interface for finding and serving documentation.
-
 %s -w <name> ...
     Write out the HTML documentation for a module to a file in the current
     directory.  If <name> contains a '%s', it is treated as a filename; if
     it names a directory, documentation is written for all the contents.
-""" % (cmd, os.sep, cmd, cmd, cmd, cmd, os.sep)
+""" % (cmd, os.sep, cmd, cmd, cmd, os.sep)
 
 if __name__ == '__main__': cli()
-- 
2.11.0

